# sample generator for embedded questions
This notebook is for creating the random sample .txt files for the experiment.

In [1]:
import pandas as pd
import numpy as np

 # Contents
 1. [Constrain the data set to stimuli set](#Constrain-the-dataset-to-stimuli-set)
 2. [Figuring out the distribution of factors per list](#Figuring-out-the-distribution-of-factors-per-list)
 3. [Figure out how to collapse the matrix verb columns](#Figure-out-how-to-collapse-the-matrix-verb-columns)
 4. [Add in the paraphrases](#Add-in-the-paraphrases)
 5. [Split EntireSentence on Question](#Split-EntireSentence-on-Question)
     2. [Wh Balancing](#Wh-Balancing)
         1. [Who](#Who)
         2. [What](#What)
         3. [Where](#Where)
         4. [When](#When)
         5. [How](#How)
 8. [Generating-random-samples](#Generating-random-samples)
     1. [First Iteration](#First-Iteration)
     2. [Second Iteration](#Second-Itreation)
     3. [Third Iteration](#Third-Iteration)
     4. [Fourth Iteration](#Fourth-Iteration)
     5. [Fifth Iteration](#Fifth-Iteration)
     6. [Sixth Iteration](#Sixth-Iteration)
     7. [Final Set](#Final-Set)

In [2]:
# import the database file from the TGrep2 searching
df = pd.read_csv("../results/swbd.tab", sep='\t', engine='python')

In [3]:
df.head()

Item_ID                                           Sentence HaveNeedTo  \
0    3:43  uh, first, um, i need *-1 to know, uh, how do ...         no   
1   17:77  and, uh, we were, i was fortunate in that i wa...         no   
2   21:45  so, i was very comfortable, you know, in *-1 d...        yes   
3   23:31  well, i had an occasion for my mother-in-law w...        yes   
4   96:22  i mean, for somebody who *t*-1 is, you know, f...         no   

  Finite ModalPresent QuestionType EmbeddedSQ DegreeQ SubjectAuxInv WhAll  \
0    yes           no     embedded        yes      no           yes   NaN   
1    yes           no      subject        NaN      no           NaN   NaN   
2    yes           no   embadjunct        NaN      no           NaN   NaN   
3    NaN          yes     relative        NaN      no           NaN   NaN   
4    yes           no     relative        NaN      no           NaN   NaN   

   ... FullWhPhrase                                   JustMatrixClause  \
0  ...          NaN  *-1 to know, uh, how do you feel *t*-2 about, ...   
1  ...          who                                                NaN   
2  ...         when  *-1 doing it when it got to the point that we ...   
3  ...          who                                                NaN   
4  ...          who                                                NaN   

  DeterminerSubjPresent DeterminerNonSubjPresent WhNode     WhParse  \
0                    no                       no    NaN   (WRB how)   
1                    no                       no    NaN    (WP who)   
2                    no                       no    NaN  (WRB when)   
3                    no                       no    NaN    (WP who)   
4                    no                       no    NaN    (WP who)   

                                            Question  \
0  how do you feel *t*-2 about, uh, about * sendi...   
1  who, uh, *t*-1 ran the nursing home in our lit...   
2  when it got to the point that we had *-2 to do...   
3  who *t*-1 had fell and needed * to be, you kno...   
4  who *t*-1 is, you know, for most of their life...   

                                       SentenceParse    WhPhaseType IdentityQ  
0  (TOP (S (INTJ (UH Uh)) (, ,) (ADVP-TMP (RB fir...  monomorphemic       NaN  
1  (TOP (S (CC And) (, ,) (INTJ (UH uh)) (, ,) (E...  monomorphemic        no  
2  (TOP (S (RB So) (, ,) (NP-SBJ-1 (PRP I)) (VP (...  monomorphemic        no  
3  (TOP (S (INTJ (UH well)) (, ,) (NP-SBJ (PRP I)...  monomorphemic        no  
4  (TOP (FRAG (PRN (S (NP-SBJ (PRP I)) (VP (VBP m...  monomorphemic        no  

[5 rows x 43 columns]

In [4]:
# This makes the display show more info
pd.set_option('display.max_rows', None)
pd.set_option('display.max_colwidth', None)

In [5]:
df.pivot_table(index=['QuestionType'], values="Question", aggfunc=len).groupby(["QuestionType"]).Question.transform(lambda x: x/len(df)).reset_index()

QuestionType  Question
0      adjunct  0.075203
1        cleft  0.064712
2   embadjunct  0.236886
3     embedded  0.160800
4     fragment  0.067654
5     relative  0.133150
6         root  0.124424
7      subject  0.137072

# Constrain the dataset to stimuli set
for experimental mock-up

First we have to remove the questions that we don;t want to include:
1. embedded questions only
2. no degree questions
3. no identity questions
4. generally only monomorphemic wh-phrases
5. only who-, what-, where-, when-, how-, and why-questions

In [6]:
critical = df[(df['QuestionType'] == 'embedded') # only embedded questions
              & 
              (df['Finite'] == 'no' )] # infinitival clauses
            

In [7]:
len(critical)

135

In [68]:
7*30

210

### Figuring out how many lists

135/20

6 lists of 20, 1 of 15

15 how / 6list
4 whats / 6 lists
then random sample 1 item from the remainders

last list is whatever's left over (total 15 items)


### Number of participants
7Lists*30 subjects/list = 210

# Combine contexts with constrained db

In [8]:
# read in df with contexts
cntxts = pd.read_csv("swbd_contexts.csv")

In [9]:
cntxts = cntxts.drop(columns="FollowingContext")

In [10]:
# get the indixes from critical
crit_index = critical.Item_ID

### Merge back in Wh colum
so we can sample proportionately based on Wh

In [13]:
df_Wh = critical[["Item_ID","Wh","Question","MatrixPredVerb","MatrixNegPresent"]].rename(columns={"Item_ID": "TGrepID"})
df_Wh

TGrepID     Wh  \
58       1049:139    how   
69        1130:56    how   
267       4531:36    how   
379       6397:23    how   
476       7981:19    who   
491       8085:15   what   
583       9927:18    how   
805      13907:24    how   
865      14735:59    how   
872      14829:73   what   
1043     18673:73   what   
1096     19512:11   what   
1145     20272:87    how   
1285     22314:16    how   
1422     24860:16   what   
1472     25772:30    how   
1473     25772:55    how   
1528     26848:18  where   
1608    28106:159    how   
1695    29430:124    how   
1716     29734:16    how   
1759     30380:92    how   
1800     31157:19    how   
1801     31157:69    how   
1828     31283:31    how   
1976     34142:42    how   
1977    34142:114    how   
2077     36073:14    how   
2081     36140:45    how   
2268     38890:31    how   
2291     39551:19   what   
2491     42920:16    how   
2492     42922:16    how   
2625     45111:93    how   
2626     45116:18    how   
2627    45129:143    how   
2628     45146:25    how   
2629     45147:85    how   
2634     45170:26    how   
2714     46762:16    how   
2715     46764:31    how   
2737     47381:11    how   
2792     48296:72    how   
2793     48297:67    how   
2797     48325:64    how   
2798     48339:27    how   
2844     49496:23    how   
2905     50809:76    how   
3047     53473:26   what   
3088     54036:19    how   
3089     54041:35    how   
3171     55459:18    how   
3226    55972:154    how   
3310     57302:38    how   
3325     57570:29  where   
3387     58645:18   what   
3390     58758:18    how   
3458    59470:107    how   
3575     61475:34    how   
3603     62072:48    how   
3617     62250:48    how   
4073     71647:47    how   
4074     71650:16    how   
4081     71768:54    how   
4099     72086:35    how   
4235    74357:139    how   
4680     81788:35    how   
4875     85292:20   what   
4940     86649:33    how   
4941     86651:16    how   
4944     86707:53    how   
4945     86712:37    how   
4947    86728:108    how   
5004     87599:18   what   
5014     87727:36    how   
5026     87941:11    how   
5027     87943:11    how   
5064     88496:36    how   
5166     90074:87   what   
5171     90180:32    how   
5216     90924:61   what   
5239     91275:50   what   
5252     91713:93    how   
5284     92015:77   what   
5551     97050:16    how   
5788    100627:16   what   
5924    103134:21   what   
5958    103781:16    how   
5978    103964:16   what   
6050    105042:18    how   
6057    105090:37   what   
6079    105552:54    how   
6111    106081:42  where   
6129    106447:53    how   
6130    106458:28   what   
6204    107569:49    how   
6299    109227:56  where   
6302    109259:18  where   
6470    112241:25    how   
6471    112247:39    how   
6479    112423:40    how   
6540    113745:51    how   
6574    114251:28    who   
6605    114731:36    how   
6606    114738:50    how   
6622    115127:13    how   
6643    115523:46    how   
6758    117586:18   what   
6810    118700:62    who   
6840    119163:33    how   
6896    120021:35    how   
6934    120636:23    how   
7007    121906:18   what   
7063    122899:21   what   
7626    133807:16    how   
7628    133817:19    how   
7979    142180:37    how   
8225    147326:16   when   
8294    148793:49    how   
8295    148810:76    how   
8396    150716:46    how   
8477    151728:33    how   
8587    153527:76    how   
8589   153560:105    how   
8622    154107:34    how   
8725    155619:56   what   
8751    155874:31   what   
9115    161808:39    how   
9240    163673:26   what   
9736    170899:22   what   
9834    172447:43    how   
9885    173250:33    how   
10134   176749:52    how   
10138   176811:47    how   
10139   176813:21    how   

                                                                                                    Question  \
58                                                         how *-5 to behave *t*-6 such t

In [14]:
# subset to the items that are just the ones filtered in the previos section

# otherwise, if using the database file with contexts directly in there, then this step
# is not necessary
df_valid = cntxts[cntxts["TGrepID"].isin(set(crit_index))]

In [16]:
# Merge
df_valid = df_valid.merge(df_Wh, how = 'inner', indicator=False)
df_valid

TGrepID  \
0      1049:139   
1       1130:56   
2       4531:36   
3       6397:23   
4       7981:19   
5       8085:15   
6       9927:18   
7      13907:24   
8      14735:59   
9      14829:73   
10     18673:73   
11     19512:11   
12     20272:87   
13     22314:16   
14     24860:16   
15     25772:30   
16     25772:55   
17     26848:18   
18    28106:159   
19    29430:124   
20     29734:16   
21     30380:92   
22     31157:19   
23     31157:69   
24     31283:31   
25     34142:42   
26    34142:114   
27     36073:14   
28     36140:45   
29     38890:31   
30     39551:19   
31     42920:16   
32     42922:16   
33     45111:93   
34     45116:18   
35    45129:143   
36     45146:25   
37     45147:85   
38     45170:26   
39     46762:16   
40     46764:31   
41     47381:11   
42     48296:72   
43     48297:67   
44     48325:64   
45     48339:27   
46     49496:23   
47     50809:76   
48     53473:26   
49     54036:19   
50     54041:35   
51     55459:18   
52    55972:154   
53     57302:38   
54     57570:29   
55     58645:18   
56     58758:18   
57    59470:107   
58     61475:34   
59     62072:48   
60     62250:48   
61     71647:47   
62     71650:16   
63     71768:54   
64     72086:35   
65    74357:139   
66     81788:35   
67     85292:20   
68     86649:33   
69     86651:16   
70     86707:53   
71     86712:37   
72    86728:108   
73     87599:18   
74     87727:36   
75     87941:11   
76     87943:11   
77     88496:36   
78     90074:87   
79     90180:32   
80     90924:61   
81     91275:50   
82     91713:93   
83     92015:77   
84     97050:16   
85    100627:16   
86    103134:21   
87    103781:16   
88    103964:16   
89    105042:18   
90    105090:37   
91    105552:54   
92    106081:42   
93    106447:53   
94    106458:28   
95    107569:49   
96    109227:56   
97    109259:18   
98    112241:25   
99    112247:39   
100   112423:40   
101   113745:51   
102   114251:28   
103   114731:36   
104   114738:50   
105   115127:13   
106   115523:46   
107   117586:18   
108   118700:62   
109   119163:33   
110   120021:35   
111   120636:23   
112   121906:18   
113   122899:21   
114   133807:16   
115   133817:19   
116   142180:37   
117   147326:16   
118   148793:49   
119   148810:76   
120   150716:46   
121   151728:33   
122   153527:76   
123  153560:105   
124   154107:34   
125   155619:56   
126   155874:31   
127   161808:39   
128   163673:26   
129   170899:22   
130   172447:43   
131   173250:33   
132   176749:52   
133   176811:47   
134   176813:21   

                                                                                                                                                                                                                           EntireSentence  \
0    i also thought about it, was of, uh, *-1 waiting *-2 to talk to you that, another thing that *t*-4 occurred to me is 0 there is not so much invasion of my privacy because i know how *-5 to behave *t*-6 such that there isn't *?*.   
1                                                                                                                                                         and if that's gone, um, i, i really don't know how *-1 to live *t*-2 very well,   
2                                                                                                                           and more and more people start *-2 believing them or wondering how * to combine them with other things *t*-1,   
3                                                                                                                                                     so, i really don't know how * to compare it to other big companies *t*-1, you know.   
4                                                                                                                                                                                      you don't even know who *-1 to payoff *t*-2, huh?.   
5

In [17]:
df_valid.head()

TGrepID  \
0  1049:139   
1   1130:56   
2   4531:36   
3   6397:23   
4   7981:19   

                                                                                                                                                                                                                         EntireSentence  \
0  i also thought about it, was of, uh, *-1 waiting *-2 to talk to you that, another thing that *t*-4 occurred to me is 0 there is not so much invasion of my privacy because i know how *-5 to behave *t*-6 such that there isn't *?*.   
1                                                                                                                                                       and if that's gone, um, i, i really don't know how *-1 to live *t*-2 very well,   
2                                                                                                                         and more and more people start *-2 believing them or wondering how * to combine them with other things *t*-1,   
3                                                                                                                                                   so, i really don't know how * to compare it to other big companies *t*-1, you know.   
4                                                                                                                                                                                    you don't even know who *-1 to payoff *t*-2, huh?.   

                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                       PreceedingContext  \
0                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                               ###speakera67.###yeah.###speakerb68.###so, maybe that is a, a little bit of what privacy is *t*-1.###speakera69.###yes.###exactly.###speakerb70.###speakera71.###uh-huh.   
1                                                                                                                                                                         ###i actually###for al-, of the time 0 i've spent *t*-1 there, i still don't quite understand how certain things that i assume 0 *t*-2 and require privacy and require not just that you be alone but actually that you have a sense of privacy *t*-3.###speakerb94.###yes.###speakera95.###because anyone can be alone for some period of time###but for me a lot of what i do *t*-1 requires a sense that there's this invisible barrier around me which people wil-, will respect *t*-2.###speakerb96.###yes.###speakera97.   
2  ###and, part of it is 0 california, you know, in, back in the sixties, had a lot of alternative movements###speakerb98.###speakera99.###and some of them fizzled out###and some of them were disastrous###and others of them, um, had an impact on the society around here.###and one of the ones that *t*-1 had an impact was, uh, people becoming interested in alternate practices,###i'm not sure if it was a meditation practice, or if it was, you know, 

# Split EntireSentence on Question 
This is necessary because we need to bold the question only and not the Matrix in the experimental file

In [76]:
# split EntireSentence at the string that matches the value in the 'question' column
df_valid["Matrix"] = df_valid.apply(lambda x: x['EntireSentence'].replace(x['Question'],"").strip(),axis=1)

In [77]:
# split that last punctuation off, to be added back on in .js script
df_valid["punct"] = df_valid["Matrix"].apply(lambda x: list(x)[-1])

In [78]:
# remove that final punct from the Matrix column
df_valid["Matrix"] = df_valid["Matrix"].apply(lambda x: x.replace(list(x)[-1], ' '))

In [79]:
df_valid.head()

TGrepID  \
0  1049:139   
1   1130:56   
2   4531:36   
3   6397:23   
4   7981:19   

                                                                                                                                                                                                                         EntireSentence  \
0  i also thought about it, was of, uh, *-1 waiting *-2 to talk to you that, another thing that *t*-4 occurred to me is 0 there is not so much invasion of my privacy because i know how *-5 to behave *t*-6 such that there isn't *?*.   
1                                                                                                                                                       and if that's gone, um, i, i really don't know how *-1 to live *t*-2 very well,   
2                                                                                                                         and more and more people start *-2 believing them or wondering how * to combine them with other things *t*-1,   
3                                                                                                                                                   so, i really don't know how * to compare it to other big companies *t*-1, you know.   
4                                                                                                                                                                                    you don't even know who *-1 to payoff *t*-2, huh?.   

                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                       PreceedingContext  \
0                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                               ###speakera67.###yeah.###speakerb68.###so, maybe that is a, a little bit of what privacy is *t*-1.###speakera69.###yes.###exactly.###speakerb70.###speakera71.###uh-huh.   
1                                                                                                                                                                         ###i actually###for al-, of the time 0 i've spent *t*-1 there, i still don't quite understand how certain things that i assume 0 *t*-2 and require privacy and require not just that you be alone but actually that you have a sense of privacy *t*-3.###speakerb94.###yes.###speakera95.###because anyone can be alone for some period of time###but for me a lot of what i do *t*-1 requires a sense that there's this invisible barrier around me which people wil-, will respect *t*-2.###speakerb96.###yes.###speakera97.   
2  ###and, part of it is 0 california, you know, in, back in the sixties, had a lot of alternative movements###speakerb98.###speakera99.###and some of them fizzled out###and some of them were disastrous###and others of them, um, had an impact on the society around here.###and one of the ones that *t*-1 had an impact was, uh, people becoming interested in alternate practices,###i'm not sure if it was a meditation practice, or if it was, you know, 

# Generating random samples

## First Iteration

6 lists of 20, 1 of 15

15 how / 6 list
4 whats / 6 lists
then random sample 1 item from the remainders

last list is whatever's left over (total 15 items)

In [80]:
#for n in range(1,7):
    how_sample = df_valid[df_valid["Wh"] == "how"].sample(15)
    df_valid = df_valid.drop(how_sample.index)    
    
    what_sample = df_valid[df_valid["Wh"] == "what"].sample(4)
    df_valid = df_valid.drop(what_sample.index)
    
    final_sample = df_valid[
        (df_valid["Wh"] != "how") &
        (df_valid["Wh"] != "what")
    ].sample(1)
    df_valid = df_valid.drop(final_sample.index)

    
    #total = pd.concat([how_sample,what_sample,final_sample])
    # save to file
    #filename = f"../../experiments/clean_corpus/01_experiment/corpus_{n}.txt".format(n=n)
    #total.to_csv(filename,header=True,sep="\t",index=False)
    
    
    
    

## Final Set

List = 15

10 how

In [81]:
len(df_valid)

15